### ISSUE TITLE
Add `-Property` parameter to the `switch` statement

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#24407]
### Summary of the new feature / enhancement

PowerShell is about `PSCustomObject`s, yet one of the first gotha's for a novice scripter is [how to enumerating its properties](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-pscustomobject#enumerating-property-names), where:
* `$myObject | Get-Member -MemberType NoteProperty | Select -ExpandProperty Name`
  is a quiet outdate syntax and might get rather slow
* `$myobject.psobject.properties.name`
  isn't obvious

Besides, both statements don't directly link to the related property `Value`.

There are many StackOverflow issues about this (see e.g.: [Iterate over PSObject properties in PowerShell](https://stackoverflow.com/questions/37688708/iterate-over-psobject-properties-in-powershell)) which concern both **custom** (e.g. the default results of `ConvertFrom-Json`) and **Component** objects.

### Proposed technical implementation details (optional)

By adding a `-Property` parameter to the `Switch` statement might lead to a better idiomatic syntax and ease the `name`/`value` relation for certain iterations.

There are of cause a lot of other syntaxes to do this but if you currently would use the `Switch` statement for this, it would look like this:

```PowerShell
foreach ($LocalUser in Get-LocalUser) {
    Switch ($LocalUser.PSObject.Properties) {
        { $_.Name -eq 'Enabled' }         { if (-not $_.Value) { Write-Host 'Action on disabled account'  } }
        { $_.Name -eq 'PasswordExpires' } { Write-Host 'Check/action upon expire date' $_.Value }
        { $_.Name -eq 'LastLogon' }       { Write-Host 'Check/action upon usage' $_.Value }
    }
}
```

**Wishful thinking**
For the suggested `-Property` parameter, the current `<result-to-be-matched>` side would compare against the `<test-expression>.PSObject.Properties.Name` and the current object (`$_`) at the `<action>` side will hold the `<test-expression>.PSObject.Properties.Value`:

```PowerShell
Switch -Property (Get-LocalUser) {
    Enabled         { if (-not $_) { Write-Host 'Action on disabled account' } }
    PasswordExpires { Write-Host 'Check/action upon expire date' $_ }
    LastLogon       { Write-Host 'Check/action upon usage' $_ }
}
```

## Addition 2024-10-12

IMO there isn't any restriction to apply this parameter thought also to dictionaries, one general thing though, is that in the action block, you don't have direct access to the other (sibling) properties in this design idea. As an example you might want to generate an error with the `LocalUser`'s name in the `Enabled` action. One might over come this by capturing the name first, e.g.:

```PowerShell
Switch -Property (Get-LocalUser) {
    Name            { $Name = $_ }
    Enabled         { if (-not $_) { Write-Error 'User $Name is disabled } }
    PasswordExpires { Write-Host 'Check/action upon expire date' $_ }
    LastLogon       { Write-Host 'Check/action upon usage' $_ }
}
```

### 2<sup>nd</sup> thought: `-Name` parameter

Or have a slightly different the design where the action block holds the complete object rather than just the value (or the `current` enumeration in case of a dictionary), which will result in to something like:
(In this design option I would call the proposed parameter something like `-Name`)

 ```PowerShell
Switch -Name (Get-LocalUser) {
    Enabled         { if (-not $_.Enabled) { Write-Error 'User $_.Name is disabled } }
    PasswordExpires { Write-Host 'Check/action upon expire date' $_.PasswordExpire }
    LastLogon       { Write-Host 'Check/action upon usage' $_.LastLogon }
}
```

### 3<sup>rd</sup> thought: `-Target (<target>)` parameter

An even more general approach could be a `-Target <name>` parameter which would allow to target any property (/member, including methods):

 ```PowerShell
Switch ((Get-LocalUser).PSObject.Properties) -Target ('Name') {
    Enabled         { if (-not $_.Value.Enabled) { Write-Host 'Action on disabled account'} }
    PasswordExpires { Write-Host 'Check/action upon expire date' $_.Value.PasswordExpire }
    LastLogon       { Write-Host 'Check/action upon usage' $_.Value.LastLogon }
}
```

This could even cover for any unforeseen target as in #17456 (where by no means I am trying to replace the specific `-Type` proposal):

```PowerShell
switch ($InputObject) -Target ('GetType') {
    string    { "I am a string! My value is: '$_'" }
    hashtable { "I am a hash table! My value is: '$($_.Message)'" }
    default   { "I am a $($_.GetType())! I'm not currently handled. :(" }
}
```

Besides, if the `-Target` parameter would also supports a `ScriptBlock` which the current object (`$_`) in scope, the target possibility will probably endless:

 ```PowerShell
switch (Get-LocalUser) -Target ({ $_.PSObject.Properties } ) { # Or possibly, just: -Target { $_.PSObject.Properties } 
    Enabled         { if (-not $_.Enabled) { Write-Host 'Action on disabled account'} }
    PasswordExpires { Write-Host 'Check/action upon expire date' $_.PasswordExpire }
    LastLogon       { Write-Host 'Check/action upon usage' $_.LastLogon }
}
```

---

Anyways, the general idea of this proposal is to **switch** upon a `name`/`key` or different target (in the `<result-to-be-matched>`  block) and by that, minimize the `if`/`else` statements required which is afaik one of the reasons to use the `switch` statement that currently only works with (a *list* of) scalars.

**Comment by rhubarb-geek-nz** [comment]
The standard approach to OO languages that don't meet your needs is first see if you can add types and classes in the style of the existing language before you say you have to change the language.

Also the "Clean Code" movement frowns on the `switch` statement, not as badly as `goto` but it isn't invited to any parties. They suggest handlers should be in dictionaries or hashtables that can be extended without cluttering up the main code.

It is good you have given examples of cmdlets that you think have shorting comings and have a proposal.

Unfortunately your proposal changes the language so I am going to propose removing one space character.

`Switch -Property` becomes `Switch-Property`

Then we can solve it with the existing PowerShell language and system.

This prototype is in PowerShell but I would expect this would really be written in C#.

```
function Switch-Property
{
    param(
        [Parameter(ValueFromPipeline=$True)][PSObject[]]$InputObject,
        [System.Collections.IDictionary]$Case
    )
}
```

The $Case parameter is actually a dictionary of arrays of `ScriptBlock`s.

We could then rewrite your code as

```
foreach ($LocalUser in Get-LocalUser) {
    Switch-Property $LocalUser @{
        Enabled = ( { -not $_ }, { Write-Host 'Action on disabled account' } )
        PasswordExpires = ( { $True }, { Write-Host 'Check/action upon expire date' $_ } )
        LastLogon = ( { $True }, { Write-Host 'Check/action upon usage' $_ } )
    }
}
```

And if we had our act together with the pipeline and inputobject we could do

```
Get-LocalUser | Switch-Property @{
        Enabled = ( { -not $_ }, { Write-Host 'Action on disabled account' } )
        PasswordExpires = ( { $True }, { Write-Host 'Check/action upon expire date' $_ } )
        LastLogon = ( { $True }, { Write-Host 'Check/action upon usage' $_ } )
}
```

And to keep the "Clean Code" people happy

```
$LocalUserHandlers = @{
        Enabled = ( { -not $_ }, { Write-Host 'Action on disabled account' } )
        PasswordExpires = ( { $True }, { Write-Host 'Check/action upon expire date' $_ } )
        LastLogon = ( { $True }, { Write-Host 'Check/action upon usage' $_ } )
}

Get-LocalUser | Switch-Property $LocalUserHandlers
```

Now as a separate cmdlet you can sort out the details, so should the $_ sent to the comparison ScriptBlock be the property, but should the $_ given to the action ScriptBlock should be the whole object?

You could also decide what happens if the array of scriptblocks is not two, is it for more conditions or for more actions?

**Comment by iRon7** [comment]
@rhubarb-geek-nz,

Thank you for your suggestion, but be aware that your prototype doesn't result in the same syntax as in the proposal.
Besides it doesn't cover for the expected behavior that goes along with embedding a feature requested into PowerShell, such as:
* The fact that it requires some kind of installation.
* Performance (due to [function costs](https://github.com/PowerShell/PowerShell/issues/19318) and [wrapping cmdlets](https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations#avoid-wrapping-cmdlet-pipelines))

Anyways, as this is your idea, I would suggest to add it to the PowerShell Gallery *yourself* and/or add it as an answer with text and explanation to (any of) the StackOverflow questions (as [the one mentioned in the inital propose](https://stackoverflow.com/questions/37688708/iterate-over-psobject-properties-in-powershell)) knowing that certain environments might not have access to the PowerShell Gallery or even require signed scripts/cmdlets.

The point I am trying to make here:

PowerShell Developers that initiate a ...

> [**Feature Request / Idea**](https://github.com/PowerShell/PowerShell/issues/new/choose) 🚀
> Suggest a new feature or improvement (this does not mean you have to implement it)

... do that with the intention to get something  changed or *integrated* into Powershell and  *not* because they looking for a DYI alternative.

> This prototype is in PowerShell but I would expect this would really be written in `C#`.

Is in fact one other argument why I think this should be considered to be covered by the PowerShell engine as I (and many other PowerShell scripters) do not have (and do not need) `C#` knowledge .

It is good to get feedback on a proposal (knowing anyone might have missed a obvious solution) but I assume (and sincerely hope) that your statement:

> The standard approach to OO languages that don't meet your needs is first see if you can add types and classes in the style of the existing language before you say you have to change the language.

isn't conform the open-source idea and being shared by the PowerShell team.
(If you want to further argue on your statement, I recommend you to open a [discussion](https://github.com/PowerShell/PowerShell/discussions/new/choose) and not further clutter this proposal)

Bottom line:
If the PowerShell team decides to reject this proposal as redundant or not conform the general PowerShell idea, I will simply accept that.

**Comment by iRon7** [comment]
I have created a prototype for the 3<sup>rd</sup> thought:

<details>
  <summary>CustomSwitch</summary>
  
  ```powershell
function Switch-Target ($Condition, [ScriptBlock]$Action) {
    $Values = 
        if ($Target -is [ScriptBlock]) {
            $Item | ForEach-Object $Target # Utilize $_
        }
        else {
            $Member = $Item.PSObject.Members[$Target]
            Switch ($Member.MemberType) {
                Method   { $Member.Invoke() }
                Property { $Member.Value }
                Default  { Throw "Unkown member: $Target" }
            }
        }
    foreach ($Value in $Values) {
        $IsEqual = if ($Condition -is [ScriptBlock]) { $Value | ForEach-Object $Condition } else { $Condition -eq $Value }
        if ($IsEqual) { $Item | ForEach-Object $Action }
    }
}

function CustomSwitch ($List, $Target, [ScriptBlock]$ScriptBlock) {
    $Ast = [System.Management.Automation.Language.Parser]::ParseInput($ScriptBlock, [ref]$Null, [ref]$Null)
    $Expressions = $Ast.EndBlock.Statements.Pipelineelements.Extent.Text 
    foreach ($Item in $List) {
        foreach ($Expression in $Expressions) { Invoke-Expression "Switch-Target $Expression" }
    }
}
  ```
</details>

To better examen the results.

It appears that although the **`-Target (<target>)` parameter** might indeed be a general parameter to target anything unforeseen member as the `-Type` switch in #17456:

```PowerShell
CustomSwitch ('Test', @{ Message = 'Hello World' }, 123) -Target 'GetType' {
    string    { "I am a string! My value is: '$_'" }
    hashtable { "I am a hash table! My value is: '$($_.Message)'" }
    default   { "I am a $($_.GetType())! I'm not currently handled. :(" } # Not implemented in prototype
}

I am a string! My value is: 'Test'
I am a hash table! My value is: 'Hello World'
```
Yet, I can't think of any other use-case for this.

Which regards to targeting properties (or dictionary items), there appear two type of use-cases for the `<result-to-be-matched>` block:
* Check if the property (or item) *name* exists (the `Get-LocalUser` isn't a good example for this, but the results from e.g. `ConvertFrom-Json` might some how leverage from a `switch` statement).
* Holding each property (or item) *value* against a particular condition

I have not been able to find a clean syntax to cover both use-cases.
Besides, it hard to define whether the `<action>` block should have an easy access to the current `<test-expression>` object or each concerned property/item (`.value`) knowing that there is only one current object (`$_`).

Therefore I have decided to close this feature request.
